Go sequence interface/struct hierarchy refactor

kalman opened the pull request:
This makes a number of changes to simplify code:
- metaSequence is now a struct, not an interface
- orderedMetaSequence and indexedMetaSequence is gone
- metaSequenceObject is gone
- indexedSequence is gone
- add leafSequence struct for leaf sequences to embed

kalman commented:
jenkins: perf this

kalman commented:
I took your patch, fixed the failing test, added leafSequence, and reverted JS changes.

I will verify that this is possible in JS, but I wouldn't want to wait for me to implement it in JS before landing the Go side. I imagine this will rot pretty quickly.

rafael-atticlabs commented on go/types/sequence_cursor.go:
```diff
@@ -19,10 +19,13 @@ func newSequenceCursor(parent *sequenceCursor, seq sequence, idx int) *sequenceC
 
-	return &sequenceCursor{parent, seq, idx}
+	cur := &sequenceCursor{parent, seq, idx}
+	d.Chk.NotNil(cur.seq)
```
did you do this or did I? also, in either case, why no check for nil above at the beginning?

kalman replied:
Ah yeah this is leftover debugging for the failing test

rafael-atticlabs commented on go/types/sequence_cursor.go:
```diff
@@ -19,10 +19,13 @@ func newSequenceCursor(parent *sequenceCursor, seq sequence, idx int) *sequenceC
 }
 
 func (cur *sequenceCursor) length() int {
+	d.Chk.NotNil(cur.seq)
```
does this buy anything? The next line will fault if it's nil, right?

kalman replied:
as is this

Commit:
Go sequence interface/struct hierarchy refactor

This makes a number of changes to simplify code:
1. metaSequence is now a struct, not an interface
2. orderedMetaSequence and indexedMetaSequence is gone
3. metaSequenceObject is gone
4. indexedSequence is gone
5. add leafSequence struct for leaf sequences to embed

Everything but change 5 was done by rafael@atticlabs.io.
